polynote-frontend/polynote/data/edit_buffer.ts

```diff
@@ -1,38 +1,46 @@
-import {NotebookUpdate} from "./messages";
+import {NotebookUpdate, UpdateCell} from "./messages";
+import * as messages from "./messages";
+import {ContentEdit} from "./content_edit";
 
 interface Version {
     version: number,
-    edits: NotebookUpdate[]
+    edit: NotebookUpdate
 }
 
 // An immutable holder of edits.
 // The old EditBuffer used to be mutable, maybe this deserves a better name now that it is immutable.
 export class EditBuffer {
 
-    constructor(readonly versions: Version[] = []) {}
+    constructor(private _versions: Version[] = []) {}
+
+    get versions(): Version[] {
+        return [...this._versions]
+    }
+
+    get duplicate(): EditBuffer {
```

Comment:
I added this method to get a duplicate of an `EditBuffer`, because I made `EditBuffer` mutable.

I did that because I couldn't see any point at which it's shared, and making it immutable means a lot of array copying (e.g. on every keystroke) which is kind of unnecessary. This also makes the logic more similar to the server-side implementation.